src/util/c_typecast.cpp

```diff
@@ -599,7 +599,60 @@ void c_typecastt::implicit_typecast_followed(
   const typet &src_type,
   const typet &dest_type)
 {
-  if (dest_type.id() == "pointer")
+  if (
+    (is_reference(dest_type) || is_rvalue_reference(dest_type)) &&
+    !(is_reference(src_type) || is_rvalue_reference(src_type)))
+  {
+    // if a destination type is an lvalue reference,
+    // which has been modelled as a pointer, but
+    // source type is not a pointer, then we just need to take
+    // the address of the source. e.g. the return statement
+    // of a function that returns an lvalue reference:
+    //    T& F(T& a) { return a; }
+    // is converted to `return &a;` rather than `return (int *)a;`
+    // See TC: regression/esbmc-cpp/template/template_1/3
+
+    // got to make sure the subtypes match each other
+    // to prevent anything that may surprise us.
+
+    if (
+      ns.follow(dest_type.subtype()).id() ==
+      src_type.id()) // TODO: remove this condition
+    {
+      address_of_exprt addr(expr);
+      addr.location() = expr.location();
+      addr.type().set("#reference", true);
+      addr.type().set("#lvalue", is_reference(src_type));
+      expr.swap(addr);
+    }
+    else
+    {
+      log_debug(
+        "c-typecast",
+        "implicit typecast of non-reference to incompatible reference type?");
+    }
+  }
+  else if (
+    (is_reference(src_type) || is_rvalue_reference(src_type)) &&
+    !(is_reference(dest_type) || is_rvalue_reference(dest_type)))
+  {
+    log_debug("c-typecast", "src-ref to non-ref dest");
+    if (ns.follow(src_type.subtype()).id() == dest_type.id())
+    {
+      dereference_exprt deref(expr, expr.type());
+      deref.location() = expr.location();
+      deref.set("#lvalue", true);
+      deref.set("#implicit", true);
+      expr.swap(deref);
+    }
+    else
+    {
+      log_debug(
+        "c-typecast",
+        "implicit typecast of reference to incompatible non-reference type?");
+    }
+  }
+  else if (dest_type.id() == "pointer")
```

Comment:
We used the `else if` here, which allows us to skip the subsequent adjustments, which I think is why we encountered `ERROR: Can't construct rvalue reference to array type during dereference (It isn't allowed by C anyway)`.

Reply:
Until 26d5b85b9eeea40ab23b106c00895c4724524ef3 esbmc did not appear this error, it affects some OM TCs at present, for example `stack`.